sklearn/utils/estimator_checks.py

```diff
@@ -2384,6 +2385,71 @@ def check_classifiers_regression_target(name, estimator_orig):
         assert_raises_regex(ValueError, msg, e.fit, X, y)
 
 
+@ignore_warnings(category=(DeprecationWarning, FutureWarning))
+def check_estimator_sparse_dense(name, estimator_orig):
+    rng = np.random.RandomState(52)
+    estimator = clone(estimator_orig)
+    estimator_sp = clone(estimator_orig)
+    tags = _safe_tags(estimator_orig)
+    centers = 2 if tags["binary_only"] else None
+    X, y = make_blobs(random_state=rng, cluster_std=0.5, centers=centers)
+    # for put some points to zero to have a little bit of sparsity
+    X -= np.min(X) - 1.  # we need positive data for ComplementNB for instance
+    X_csr = sparse.csr_matrix(X)
+    if tags["multioutput"] or tags["multioutput_only"]:
+        y = y[:, np.newaxis]
+
+    for sparse_format in ['csr', 'csc', 'dok', 'lil', 'coo', 'dia', 'bsr']:
+        X_sp = X_csr.asformat(sparse_format)
+        # catch deprecation warnings
+        with ignore_warnings(category=DeprecationWarning):
+            if name in ['Scaler', 'StandardScaler']:
+                estimator.set_params(with_mean=False)
+                estimator_sp.set_params(with_mean=False)
+            if name in ['RobustScaler']:
+                # XXX naming is not consistent (with_mean vs with_centering) :(
+                estimator.set_params(with_centering=False)
+                estimator_sp.set_params(with_centering=False)
+
+        set_random_state(estimator)
+        set_random_state(estimator_sp)
+        X = pairwise_estimator_convert_X(X, estimator)
+        X_sp = pairwise_estimator_convert_X(X_sp, estimator)
+        try:
+            with ignore_warnings(category=DeprecationWarning):
+                estimator_sp.fit(X_sp, y)
+                estimator.fit(X, y)
+            if hasattr(estimator, "predict"):
+                pred = estimator.predict(X)
+                pred_sp = estimator_sp.predict(X_sp)
+                if not (isinstance(estimator, BaseSGD) or  # SGD is too random
+                        getattr(estimator, 'kernel', None) == 'precomputed' or
+                        getattr(estimator, 'metric', None) == 'precomputed'):
+                    assert_allclose(pred, pred_sp)
+
+                assert pred.shape == pred_sp.shape
+            if hasattr(estimator, 'predict_proba'):
+                probs = estimator.predict_proba(X)
+
+                if isinstance(probs, list):
+                    # XXX : hack for dummy classifier
+                    probs = probs[0]
+
+                assert probs.shape == (X.shape[0], len(np.unique(y)))
+        except TypeError as e:
+            if 'sparse' not in str.lower(repr(e)):
+                print("Estimator %s doesn't seem to fail gracefully on "
```

Comment:
These and the below cases are not covered by tests. I think we should create a tests with a simple estimator to check this branch.

Also print goes to stdout while raised exceptions are shown on stderr (I think). That could lead to subtitle issues, but I'm not sure if it's worth addressing.

Reply:
I modified an already existing test to cover this (see comments below)

Regarding the print, there is some pre-existing explanation in the test (copied below):
```
    # the check for sparse input handling prints to the stdout,
    # instead of raising an error, so as not to remove the original traceback.
    # that means we need to jump through some hoops to catch it.
```